Enhance: get and set state of push-down-buttons (button love part1)

SlySven opened the pull request:
#### Brief overview of PR changes/additions
Provide a means to get and set the state of push down buttons (the latter without executing the associate command or script). It requires adding members to the `TAction` class to keep track of the `EAction` or `TFlipButton` instance that is the implementation of the button on a `TEasyButton` or `TToolBar` respectively. As it happens this showed that the previous code to regenerate them was creating new buttons each time but was not deleting the existing ones that were getting replaced.

Also:
* made some members of the `TAtion` class (the name, script and both commands) private as otherwise there is no point in the setters and getters also provided!
* simplified the existing `getButtonState()` {that only works within the script for a button) to remove an unneeded intermediate local (`int`) value.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

#### Motivation for adding to Mudlet
It is something has been requested in the past (I think recently on Discord). 

#### Other info (issues closed, discussion etc)
Further improvements to buttons/menus/toolbars are desirable and I can foresee some of them.

#### Release post highlight
Allow Lua scripts to get the state of *click-down* (toggle-able) buttons outside of the script for said button and allow for them to also set that state.

SlySven committed:
Enhance: add Push-down Button state getters and setters (Button Luv Part 1)

Provide a means to get and set the state of push down buttons (the latter
without executing the associate command or script). It requires adding
members to the `TAction` class to keep track of the `EAction` or
`TFlipButton` instance that is the implimentation of the button on a
`TEasyButton` or `TToolBar` respectively. As it happens this showed that
the previous code to regenerate them was creating new buttons each time but
was not deleting the existing ones that were getting replaced.

Also:
* made some members of the `TAtion` class (the name, script and both
commands) private as otherwise there is no point in the setters and getters
also provided!
* simpliedied the existing `getButtonState()` {that only works within the
script for a button) to remove an unneeded intermediate local (`int`)
value.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Tweak: I forgot to copy some later code to the first setter

The original comment was wrong as well - using a `Q_ASSERT_X` would have
broken things for a floatable/dockable toolbar type button (what
`TAction::mpFButton` will point at)...!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -13830,6 +13830,10 @@ void TLuaInterpreter::initLuaGlobals()
     lua_register(pGlobalLua, "getProfileTabNumber", TLuaInterpreter::getProfileTabNumber);
     lua_register(pGlobalLua, "addFileWatch", TLuaInterpreter::addFileWatch);
     lua_register(pGlobalLua, "removeFileWatch", TLuaInterpreter::removeFileWatch);
+    lua_register(pGlobalLua, "getButtonStateById", TLuaInterpreter::getButtonStateById);
+    lua_register(pGlobalLua, "getButtonStateByName", TLuaInterpreter::getButtonStateByName);
+    lua_register(pGlobalLua, "setButtonStateById", TLuaInterpreter::setButtonStateById);
+    lua_register(pGlobalLua, "setButtonStateByName", TLuaInterpreter::setButtonStateByName);
```
Can you merge these so it's consistent with the rest of the Lua API, ie using either name or id works?

SlySven replied:
Um, not sure that is wise - since the name of the button is currently ALSO the text that is displayed (I have plans to separate that) - what happens if the name of the button **is** a number? That is a usage that is not *totally* unreasonable from a UI point of view, *I* think?

🤔 I suppose that one could check the `lua_type(...)` of the provided argument (instead of using `lua_isnumber(...)` or `lua_isstring(...)` which will (un)happily (in this case) *coerce* from a string that is a number to that number and vice-versa). So that it has to be a number to be treated as an ID.

Note that we already have stupid implementations in some functions where an ID *number* is required but it must be provided as a **string**!

SlySven replied:
Ah, I have found that I *can* combine the new getter functionality within the existing `getButtonState()` function - use it without arguments for the existing (only within the script for the button concerned) or provide a *number* for an ID or a *"string"* for the (last one registered with that string as it's) name. The setters can be similarly combined as well.

Commits to do this will follow shortly - expected by the end of the day...

SlySven committed:
Revise: combine ID and name getters and setters for button state

Note that for the getter this is an intermediate stage as is should be
possibble to combine it with the existing `getButtonState()` function that
takes NO arguments but only works within the Lua script for a push-down
button.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: combine new getter functionality into existing getButtonState

This simpilifies the UI so that `getButtonState()` works the same as
before but if an button's ID number or "name" is supplied as an argument
the state of the corresponding button (as a boolean though, NOT the
existing and, TBH, odd, integer value 1 or 2, that the original function
produces) is returned. In the event that the matching item is NOT a
push-down button, or that it does not match an existing button then a 'nil'
+ (run-time) error message is returned instead.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

keneanung commented on src/TLuaInterpreter.cpp:
```diff
@@ -5778,10 +5780,114 @@ int TLuaInterpreter::sendTelnetChannel102(lua_State* L)
+    }
+
+    // Otherwise consider the type of the first argument
+    auto argType = lua_type(L, 1);
+    TAction* pItem = nullptr;
+    if (argType == LUA_TNUMBER) {
+        int id = lua_tonumber(L, 1);
+        if (id < 0) {
+            return warnArgumentValue(L, __func__, QStringLiteral("item Id (%1) invalid, it must be equal or greater than zero").arg(id).toUtf8().constData());
+        }
+        pItem = host.getActionUnit()->getAction(id);
+        if (!pItem) {
+            return warnArgumentValue(L, __func__, QStringLiteral("no button item with Id %1 found").arg(id).toUtf8().constData());
+        }
+        if (!pItem->isPushDownButton()) {
+            return warnArgumentValue(L, __func__, QStringLiteral("item Id with %1 is not a push-down button").arg(id).toUtf8().constData());
+        }
+    }
+
+    if (argType == LUA_TSTRING) {
+        QString name = lua_tostring(L, 1);
+        if (name.isEmpty()) {
+            return warnArgumentValue(L, __func__, "item name must not be an empty string");
+        }
+        pItem = host.getActionUnit()->findAction(name);
+        if (!pItem) {
+            return warnArgumentValue(L, __func__, QStringLiteral("no button item with name '%1' found").arg(name).toUtf8().constData());
+        }
+        if (!pItem->isPushDownButton()) {
+            return warnArgumentValue(L, __func__, QStringLiteral("item with name '%1' is not a push-down button").arg(name).toUtf8().constData());
+        }
+    }
```
This whole part seems to be pretty much the same as in the set function. I would suggest putting this into a function that's used by both.

SlySven replied:
TBH if the code was to be used in *more* than these *two* cases I'd go along with that idea but the additional complexity of a further function call for just this pair means I think it is a borderline decision and I'd prefer not to do that currently...

SlySven replied:
:thinking: 

In the medium term I'd like to have a set of *introspection* (?) functions that can *get*, and for some *set*, the common elements of all the mudlet item types (alias, button/menu/toolbars, keybindings, scripts, timers, triggers) i.e. "name", "command", "ID", "active" {based on both it's own *and* its ancestors' state} and possibly some further details... as such it may then be useful to factor out item identification based on EITHER a numeric ID OR a string name basis. At that point this would be useful to revisit.

keneanung replied:
To me, this is not a borderline decision. It has strings and quite a bit of logic that should stay the same. And the easiest way for that is adding a function.

keneanung reviewed (commented):
One suggestion for me, but otherwise I like this.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -5778,10 +5780,114 @@ int TLuaInterpreter::sendTelnetChannel102(lua_State* L)
+            return warnArgumentValue(L, __func__, QStringLiteral("no button item with Id %1 found").arg(id).toUtf8().constData());
+        }
+        if (!pItem->isPushDownButton()) {
+            return warnArgumentValue(L, __func__, QStringLiteral("item Id with %1 is not a push-down button").arg(id).toUtf8().constData());
```
```suggestion
            return warnArgumentValue(L, __func__, QStringLiteral("item with Id %1 is not a push-down button").arg(id).toUtf8().constData());
```

we discussed earlier if we go for "id" or "Id" or "ID" and can't remember the consensus reached.. but I feel like "Id" is the worst of those ☹️

SlySven replied:
Um, one reference might be: https://github.com/Mudlet/Mudlet/issues/2605 - *and* it has not been closed...

Tomayto, Tomaahtoe ....

SlySven replied:
Ah, I'll revise it to `ID` and hope that'll silence the critics... :crossed_fingers:

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -5778,10 +5780,114 @@ int TLuaInterpreter::sendTelnetChannel102(lua_State* L)
+
+    if (!pItem) {
+        // we'll get here if the (first) argument is NOT usable:
+        lua_pushfstring(L, "%s: bad argument #1 type (ID as number or name as string expected, got %s!)",
```
See there is an "ID", let's use that consistently, shall we?

SlySven replied:
Okay, in the scheme of things I can't be ~~arsed~~*bothered* to fight this...

Kebap reviewed (commented):
* [ ] setButtonState is missing in Area 51

> set the state of push down buttons (without executing the associate command or script). 

Is that really always desired? Maybe a toggle would make sense. Not sure if this should even be default. Then again, I hardly used buttons myself.

SlySven commented:
>> set the state of push down buttons (without executing the associate command or script).
>>
>Is that really always desired? Maybe a toggle would make sense. Not sure if this should even be default. Then again, I hardly used buttons myself.

When configuring a UI that uses buttons/menus/toolbars it is pretty much essential to be able to set the state of push-down buttons to a required initial or to reflect something that is happening in-game independently of whatever script/command is associated with that button. A bit of refactoring to make the Lua script a simple function call to a script defined elsewhere (in a script element I guess) would be enough to reproduce clicking the button whereas only providing a function to simulate toggling the button state AND executing the script / send the command is actually a much less flexible way of doing things - and would not give the same scope for usage that the setter function herein gives...

SlySven committed:
Revise: used ID consistently in two functions added/modified in this PR

As suggested by peer-review.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Refactor: extract code to get TAction instance from {g|s}etButtonState(...)

This was strongly requested by @keneanung though my preference was against
it for just two instances - as developing an API that would work with the
`warnArgumentValue(...)` function took a bit of thought and I had to work
with handling a `TAction**` (pointer to a pointer to a `TAction`) as a
function argument!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -5775,13 +5777,104 @@ int TLuaInterpreter::sendTelnetChannel102(lua_State* L)
+
+    auto checked = getVerifiedBool(L, __func__, 2, "checked");
+
+    if (pItem->mButtonState != checked) {
```
:-1: **github-actions**

Both of these are false positives because they do not recognise that this does not actually return execution to the caller function (but instead to the point where `setjmp(...)` was called):
>```cpp
>src/TLuaInterpreter.cpp:5827:9: note: Returning value, which participates in a condition later
>        return lua_error(L); // Does not return!
>````

keneanung commented on src/TLuaInterpreter.cpp:
```diff
@@ -5775,13 +5777,104 @@ int TLuaInterpreter::sendTelnetChannel102(lua_State* L)
+//   it also resets the pItem to be a nullptr
+// does NOT return (normally) if the index item on the stack is not a number or
+//   a string
+int TLuaInterpreter::getTActionFromIdOrName(lua_State* L, TAction** pItem, const int index, const char* func)
```
I know this isn't atypical for C++, but I prefer using function arguments as input instead of output. This is also mentioned in Clean Code. Have you considered returning multiple values via https://www.geeksforgeeks.org/returning-multiple-values-from-a-function-using-tuple-and-pair-in-c/?

vadi2 replied:
Agree - and we have many examples of this, check for example:

```cpp
    std::pair<bool, QString> changeModuleSync(const QString& enableModuleName, const QLatin1String &value);
    std::pair<bool, QString> getModuleSync(const QString& moduleName);
```

keneanung reviewed (approved):
Thank you very much for the rework :smile: I have a niggle that I'd love to see a change for. But it won't break my heart if you're adamant here :wink:

SlySven committed:
Revise: use a different method to return the pointer to a TAction

Make it a std::pair with the original return value of
`TLuaInterpreter::getTActionFromIdOrName(...)` instead of passing it as a
non-const reference argument.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (approved):
Thanks for adding this requested feature in!

SlySven commented:
:books: documentation done, see:
* ~~**[getButtonState(...)](https://wiki.mudlet.org/w/Manual:Mudlet_Object_Functions#getButtonState)**~~ - pulled from main Wiki
* ~~**[setButtonState(...)](https://wiki.mudlet.org/w/Manual:Mudlet_Object_Functions#setButtonState)**~~ - pulled from main Wiki
* **[getButtonState(...)](https://wiki.mudlet.org/w/Area_51#getButtonState)** - relocated to Area 51
* **[setButtonState(...)](https://wiki.mudlet.org/w/Area_51#setButtonState)** - relocated to Area 51

vadi2 commented:
Looks good! But it should be in Area 51 as it's not available yet. People naturally get confused when we list releases in docs that aren't out yet

Kebap commented:
Please honor Area 51 template, e.g. don't hide version info below example section

SlySven commented:
> Looks good! But it should be in Area 51 as it's not available yet. People naturally get confused when we list releases in docs that aren't out yet

Ah, I was thinking that it being in the PTB would be justification for it making it into the main documentation.

> Please honor Area 51 template, e.g. don't hide version info below example section

🤔 I was trying to avoid contaminating the auto-completion ~~extraction process~~ *scraper*.

Anyhow I have now relocated them, in the meantime, to Area 51 and perhaps someone else could reshape the items so as to fit better...

Kebap commented:
> perhaps someone else could reshape the items so as to fit better...

someone else did

Kebap commented:
> 
> 
> 📚 documentation done, see:
> 
>     * ~**[getButtonState(...)](https://wiki.mudlet.org/w/Manual:Mudlet_Object_Functions#getButtonState)**~ - pulled from main Wiki
> 
>     * ~**[setButtonState(...)](https://wiki.mudlet.org/w/Manual:Mudlet_Object_Functions#setButtonState)**~ - pulled from main Wiki
> 
>     * **[getButtonState(...)](https://wiki.mudlet.org/w/Area_51#getButtonState)** - relocated to Area 51
> 
>     * **[setButtonState(...)](https://wiki.mudlet.org/w/Area_51#setButtonState)** - relocated to Area 51

published again for upcoming release
